ETL-439: ETL-441: tracking API and components

sudopower opened the pull request:
Events tracked in API and components.

| name | source |
|--------|--------|
| readiness_ping | api |
| pipeline_create | api |
| pipeline_stop | api |
| pipeline_edit | api |
| pipeline_terminate | api | 
| pipeline_resume | api | 
| pipeline_metrics | api | 
| ingestor_ready | ingestor | 
| sink_ready | sink | 
| join_ready | join | 
| dedup_ready | dedup | 
| ingestor_terminated | ingestor | 
| sink_terminated | sink | 
| join_terminated | join | 
| dedup_terminated | dedup | 
| ingestor_crashed | ingestor | 
| sink_crashed | sink | 
| join_crashed | join | 
| dedup_crashed | dedup |

PabloPardoGarcia committed:
add first version of transformation docs

PabloPardoGarcia committed:
refine transformations docs

sudopower committed:
feat: analytics tracking for stream metrics

sudopower committed:
fix: go fmt

sudopower committed:
feat: add tracking env vars

sudopower committed:
feat: track API readiness

sudopower committed:
feat: add tracking to create / edit pipeline

sudopower committed:
feat: init tracking client & stream metrics tracker in the background

sudopower committed:
refactor: add route from context and replica counts

sudopower committed:
feat: add route to context

sudopower committed:
chore: tidy modules

sudopower committed:
feat: init tests with tracking false

sudopower committed:
fix: unit tests missing tracking args for pipeline svc

sudopower committed:
Merge branch 'main' into sudopower/etl-439-tracking-backend-api

sudopower committed:
refactor: enable logging in usage stats client

sudopower committed:
fix: tracking URL construction

sudopower committed:
feat: add readiness ping for components

sudopower committed:
fix: check if tracking client is nil

sudopower committed:
refactor: defer tracking client context cancellation

sudopower committed:
refactor: allow nil property tracking events

nnaumov-glassflow committed:
stateless transofomations dedup integration

sudopower committed:
refactor: seperate context for async operation

sudopower committed:
refactor: remove stream and topic names from usageInfo

vladamon committed:
fix filter form wording, include unmapped filter fields

sudopower committed:
revert: health api changes

sudopower committed:
fix: typo edit pipeline tracking event name

sudopower committed:
Merge branch 'main' into sudopower/etl-439-tracking-backend-api

vladamon committed:
move filters hydration to prevent loading lag and decouple from topic

vladamon committed:
fix filter form edit mode

PabloPardoGarcia committed:
add dedup and join screenshots

vladamon committed:
add getter for server env variables

vladamon committed:
remove build time variable for api url

vladamon committed:
swith to server var for api url

nnaumov-glassflow committed:
fix float32/float64 ch clickhouse parsing

PabloPardoGarcia committed:
add dedup and postgres components

PabloPardoGarcia committed:
add filter example screenshots

PabloPardoGarcia committed:
fixes etl-568

PabloPardoGarcia committed:
fixes etl-566

nnaumov-glassflow committed:
persist tags on pipeline creation

nnaumov-glassflow committed:
swagger documentation for backend #4

nnaumov-glassflow committed:
return default empty value on empty input for transform

sudopower committed:
refactor: rename usageStats service

PabloPardoGarcia committed:
add stateless transformations

PabloPardoGarcia committed:
fix ack comments

PabloPardoGarcia committed:
update architecture diagram

sudopower committed:
fix: go fmt

PabloPardoGarcia committed:
add ui screenshot for transformation

nnaumov-glassflow committed:
support int and float for datetime conversion

nnaumov-glassflow committed:
swagger documentation for backend #5

vladamon committed:
upgrade nextjs and react to bring security patches

dependabot[bot] committed:
Bump github.com/expr-lang/expr from 1.17.6 to 1.17.7 in /glassflow-api

Bumps [github.com/expr-lang/expr](https://github.com/expr-lang/expr) from 1.17.6 to 1.17.7.
- [Release notes](https://github.com/expr-lang/expr/releases)
- [Commits](https://github.com/expr-lang/expr/compare/v1.17.6...v1.17.7)

---
updated-dependencies:
- dependency-name: github.com/expr-lang/expr
  dependency-version: 1.17.7
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

nnaumov-glassflow commented on glassflow-api/internal/service/pipeline.go:
```diff
@@ -86,6 +91,8 @@ func (p *PipelineService) CreatePipeline(ctx context.Context, cfg *models.Pipeli
 		return fmt.Errorf("insert pipeline: %w", err)
 	}
 
+	p.recordPipelineEvent(ctx, "pipeline_create", cfg.ID, cfg)
```
ah, I see that there's middleware to extract route, but seems like we don't use it?

sudopower replied:
https://github.com/glassflow/clickhouse-etl/pull/413/changes#diff-edadaa253dafe52a7c384b2b91af46a669a7bcc35d4b0450895f87a64728141fR494

nnaumov-glassflow committed:
handle datetime64 parameters in mapping

Nuzhnov committed:
ETL-483: Schema validation && ETL-538: Mappings store

sudopower committed:
fix: incorrect transformer arg sent to dedup

sudopower committed:
refactor: organize tracking package

sudopower committed:
refactor: usageStats pkg logging cleanup

sudopower committed:
refactor: remove unused context

sudopower committed:
refactor: component state event names

sudopower committed:
refactor: usagestats pkg err logging

sudopower committed:
refactor: usagestats retry mechanism

sudopower committed:
refactor: pipeline metrics collection interval

sudopower committed:
refactor: usageStats default enabled

sudopower commented on glassflow-api/internal/api/middlewares.go:
```diff
@@ -134,3 +142,17 @@ func extractRoute(r *http.Request) string {
 }
+
+// RouteContext adds the route to the request context for usage stats purposes
+func RouteContext() func(http.Handler) http.Handler {
```
This is possible but the routing paths need to be refactored, there is no efficient way to differentiate between pipeline management routes and other routes, except regex + check HTTP verbs. I would vote for first using router groups in mux and then handling this in middleware

sudopower replied:
Discussed with @nnaumov-glassflow, suggested using Huma handler registerer to intercept and use OperationID from Huma.Operation (which I had overlooked !) 

So this method works, implemented.

sudopower committed:
Merge branch 'main' into sudopower/etl-439-tracking-backend-api

# Conflicts:
#	glassflow-api/cmd/glassflow/deduplication.go

sudopower committed:
refactor: remove usage stats client from pipeline service tests

sudopower committed:
refactor: remove usage stats client from pipeline service

sudopower committed:
feat: async API event usage stats collection
Intercept pipeline ID and operation and send it to usage stats client.
The client collects it via channel and fetches the pipeline from DB and builds the event to be sent

sudopower committed:
refactor: remove unused middleware

sudopower committed:
refactor: clean up unused constants

sudopower committed:
Merge branch 'main' into sudopower/etl-439-tracking-backend-api
